{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a14f7b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools as pbt\n",
    "import pysam\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as ss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c16c0ad0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import splanl.junction_scorer as jn\n",
    "import splanl.merge_bcs as mbcs\n",
    "import splanl.coords as cds\n",
    "import splanl.plots as sp\n",
    "import splanl.score_motifs as sm\n",
    "import splanl.inspect_variants as iv\n",
    "import splanl.post_processing as pp\n",
    "import splanl.custom_splai_scores as css"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "291f5ef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr11 = '/nfs/kitzman3/lab_common/refs/human/ucsc_hg19/bychrom/chr11.fa'\n",
    "chr11_seq = pp.get_refseq( chr11 )[ 0 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b99b13a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_in = '/nfs/kitzman2/smithcat/proj/wt1_2022/ex9_data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f609c005",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide = pd.read_table( data_in + 'wt1_ex9_by_var_effects_snvs_wide-2022-0727.txt' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c150e90",
   "metadata": {},
   "outputs": [],
   "source": [
    "lit = pd.read_table( data_in + 'wt1_ex9_variant_lit.2022-0721.txt' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "01f2db09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hg19_pos</th>\n",
       "      <th>cDNA_pos</th>\n",
       "      <th>cDNA ref</th>\n",
       "      <th>cDNA alt</th>\n",
       "      <th>In ClinVar?</th>\n",
       "      <th>ClinVar interp</th>\n",
       "      <th>Refeence from ClinVar</th>\n",
       "      <th>source where variant were reported</th>\n",
       "      <th>Interpretation_ACMG_Varsome</th>\n",
       "      <th>Human disease reported in HGMD</th>\n",
       "      <th>In vitro evidence from other literatures</th>\n",
       "      <th>Reference from HGMD</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>32413697</td>\n",
       "      <td>c.1355-87</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>GeneDx Variant Classification Process</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>Benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32413620</td>\n",
       "      <td>c.1355-10</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32413620</td>\n",
       "      <td>c.1355-10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>VUS</td>\n",
       "      <td>Breast cancer?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Park (2018) BMC Cancer 18,</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>32413618</td>\n",
       "      <td>c.1355-8</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>VUS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32413617</td>\n",
       "      <td>c.1355-7</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>32413611</td>\n",
       "      <td>c.1355-1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>FSGS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kanemoto (2007) Pediatr Nephrol 22, 454</td>\n",
       "      <td>Although not reported in ClinVar, it was repor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>32413517</td>\n",
       "      <td>c.1447+1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Frasier syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Miyoshi (2006) Endocr J 53, 371</td>\n",
       "      <td>Although not reported in ClinVar, it was repor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>32413517</td>\n",
       "      <td>c.1447+1</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>FSGS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gast (2016) Nephrol Dial Transplant 31, 961</td>\n",
       "      <td>Although not reported in ClinVar, it was repor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>32413516</td>\n",
       "      <td>c.1447+2</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely pathogenic</td>\n",
       "      <td>Kikuchi (1998) J Med Genet 35, 45</td>\n",
       "      <td>ClinVar &amp; HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Frasier syndrome</td>\n",
       "      <td>Loss of +KTS isoform in minigene assay and in ...</td>\n",
       "      <td>Kikuchi (1998) J Med Genet 35, 45</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>32413515</td>\n",
       "      <td>c.1447+3</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VUS</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>32413515</td>\n",
       "      <td>c.1447+3</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kireport</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>in urine-derived patient cells shift from KTS+...</td>\n",
       "      <td>https://www.kireports.org/article/S2468-0249(2...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>32413514</td>\n",
       "      <td>c.1447+4</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>Multiple references and pulications including ...</td>\n",
       "      <td>ClinVar &amp; HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Frasier syndrome</td>\n",
       "      <td>Loss of +KTS isoform in minigene assay</td>\n",
       "      <td>Barbaux (1997) Nat Genet 17, 467</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>32413513</td>\n",
       "      <td>c.1447+5</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>Multiple references and pulications including ...</td>\n",
       "      <td>ClinVar &amp; HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Frasier syndrome</td>\n",
       "      <td>Loss of +KTS isoform in minigene assay</td>\n",
       "      <td>Bruening (1992) Nat Genet 1, 144</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>32413513</td>\n",
       "      <td>c.1447+5</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Frasier syndrome</td>\n",
       "      <td>Loss of +KTS isoform in minigene assay</td>\n",
       "      <td>Kikuchi (1998) J Med Genet 35, 45</td>\n",
       "      <td>Although not reported in ClinVar, it was repor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>32413512</td>\n",
       "      <td>c.1447+6</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>Barbaux (1997) Nat Genet 17, 467</td>\n",
       "      <td>ClinVar &amp; HGMD</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Frasier syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Barbaux (1997) Nat Genet 17, 467</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>32413511</td>\n",
       "      <td>c.1447+7</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>32413505</td>\n",
       "      <td>c.1447+13</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>VUS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>32413499</td>\n",
       "      <td>c.1447+19</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>VUS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>32413498</td>\n",
       "      <td>c.1447+20</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>Benign</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>32413528</td>\n",
       "      <td>c.1437</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>conflicting interpretations</td>\n",
       "      <td>Multiple submitters</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ambiguous genetalia (ClinVar)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>32413531</td>\n",
       "      <td>c.1434</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>32413539</td>\n",
       "      <td>c.1426</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>32413540</td>\n",
       "      <td>c.1425</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>32413543</td>\n",
       "      <td>c.1422</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>32413554</td>\n",
       "      <td>c.1411</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>32413555</td>\n",
       "      <td>c.1410</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>32413561</td>\n",
       "      <td>c.1404</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>32413564</td>\n",
       "      <td>c.1401</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>32413567</td>\n",
       "      <td>c.1398</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>32413570</td>\n",
       "      <td>c.1395</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>32413576</td>\n",
       "      <td>c.1389</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>32413594</td>\n",
       "      <td>c.1371</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Yes</td>\n",
       "      <td>likely benign</td>\n",
       "      <td>Invitae Variant Classification Sherloc</td>\n",
       "      <td>ClinVar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    hg19_pos   cDNA_pos cDNA ref cDNA alt In ClinVar?  \\\n",
       "0   32413697  c.1355-87        G        A         Yes   \n",
       "1   32413620  c.1355-10        G        C         Yes   \n",
       "2   32413620  c.1355-10        G        A          No   \n",
       "3   32413618   c.1355-8        C        T         Yes   \n",
       "4   32413617   c.1355-7        C        T         Yes   \n",
       "5   32413611   c.1355-1        G        A          No   \n",
       "6   32413517   c.1447+1        G        A          No   \n",
       "7   32413517   c.1447+1        G        C          No   \n",
       "8   32413516   c.1447+2        T        C         Yes   \n",
       "9   32413515   c.1447+3        G        A         Yes   \n",
       "10  32413515   c.1447+3        G        T          No   \n",
       "11  32413514   c.1447+4        C        T         Yes   \n",
       "12  32413513   c.1447+5        G        A         Yes   \n",
       "13  32413513   c.1447+5        G        T          No   \n",
       "14  32413512   c.1447+6        T        A         Yes   \n",
       "15  32413511   c.1447+7        A        G         Yes   \n",
       "16  32413505  c.1447+13        T        C         Yes   \n",
       "17  32413499  c.1447+19        A        G         Yes   \n",
       "18  32413498  c.1447+20        C        T         Yes   \n",
       "19  32413528     c.1437        A        G         Yes   \n",
       "20  32413531     c.1434        T        C         Yes   \n",
       "21  32413539     c.1426        A        C         Yes   \n",
       "22  32413540     c.1425        C        G         Yes   \n",
       "23  32413543     c.1422        C        T         Yes   \n",
       "24  32413554     c.1411        C        T         Yes   \n",
       "25  32413555     c.1410        C        T         Yes   \n",
       "26  32413561     c.1404        C        T         Yes   \n",
       "27  32413564     c.1401        G        A         Yes   \n",
       "28  32413567     c.1398        C        G         Yes   \n",
       "29  32413570     c.1395        C        T         Yes   \n",
       "30  32413576     c.1389        A        G         Yes   \n",
       "31  32413594     c.1371        G        A         Yes   \n",
       "\n",
       "                 ClinVar interp  \\\n",
       "0                 likely benign   \n",
       "1                 likely benign   \n",
       "2                           NaN   \n",
       "3                 likely benign   \n",
       "4                 likely benign   \n",
       "5                           NaN   \n",
       "6                           NaN   \n",
       "7                           NaN   \n",
       "8             likely pathogenic   \n",
       "9                           VUS   \n",
       "10                          NaN   \n",
       "11                   pathogenic   \n",
       "12                   pathogenic   \n",
       "13                          NaN   \n",
       "14                   pathogenic   \n",
       "15                likely benign   \n",
       "16                likely benign   \n",
       "17                likely benign   \n",
       "18                likely benign   \n",
       "19  conflicting interpretations   \n",
       "20                likely benign   \n",
       "21                likely benign   \n",
       "22                likely benign   \n",
       "23                likely benign   \n",
       "24                likely benign   \n",
       "25                likely benign   \n",
       "26                likely benign   \n",
       "27                likely benign   \n",
       "28                likely benign   \n",
       "29                likely benign   \n",
       "30                likely benign   \n",
       "31                likely benign   \n",
       "\n",
       "                                Refeence from ClinVar  \\\n",
       "0               GeneDx Variant Classification Process   \n",
       "1              Invitae Variant Classification Sherloc   \n",
       "2                                                 NaN   \n",
       "3              Invitae Variant Classification Sherloc   \n",
       "4              Invitae Variant Classification Sherloc   \n",
       "5                                                 NaN   \n",
       "6                                                 NaN   \n",
       "7                                                 NaN   \n",
       "8                   Kikuchi (1998) J Med Genet 35, 45   \n",
       "9              Invitae Variant Classification Sherloc   \n",
       "10                                                NaN   \n",
       "11  Multiple references and pulications including ...   \n",
       "12  Multiple references and pulications including ...   \n",
       "13                                                NaN   \n",
       "14                   Barbaux (1997) Nat Genet 17, 467   \n",
       "15             Invitae Variant Classification Sherloc   \n",
       "16             Invitae Variant Classification Sherloc   \n",
       "17             Invitae Variant Classification Sherloc   \n",
       "18             Invitae Variant Classification Sherloc   \n",
       "19                                Multiple submitters   \n",
       "20             Invitae Variant Classification Sherloc   \n",
       "21             Invitae Variant Classification Sherloc   \n",
       "22             Invitae Variant Classification Sherloc   \n",
       "23             Invitae Variant Classification Sherloc   \n",
       "24             Invitae Variant Classification Sherloc   \n",
       "25             Invitae Variant Classification Sherloc   \n",
       "26             Invitae Variant Classification Sherloc   \n",
       "27             Invitae Variant Classification Sherloc   \n",
       "28             Invitae Variant Classification Sherloc   \n",
       "29             Invitae Variant Classification Sherloc   \n",
       "30             Invitae Variant Classification Sherloc   \n",
       "31             Invitae Variant Classification Sherloc   \n",
       "\n",
       "   source where variant were reported Interpretation_ACMG_Varsome  \\\n",
       "0                             ClinVar                      Benign   \n",
       "1                             ClinVar               Likely benign   \n",
       "2                                HGMD                         VUS   \n",
       "3                             ClinVar                         VUS   \n",
       "4                             ClinVar               Likely benign   \n",
       "5                                HGMD                  Pathogenic   \n",
       "6                                HGMD                  Pathogenic   \n",
       "7                                HGMD                  Pathogenic   \n",
       "8                      ClinVar & HGMD                  Pathogenic   \n",
       "9                             ClinVar                  Pathogenic   \n",
       "10                           Kireport                  Pathogenic   \n",
       "11                     ClinVar & HGMD                  Pathogenic   \n",
       "12                     ClinVar & HGMD                  Pathogenic   \n",
       "13                               HGMD                  Pathogenic   \n",
       "14                     ClinVar & HGMD                  Pathogenic   \n",
       "15                            ClinVar                  Pathogenic   \n",
       "16                            ClinVar                         VUS   \n",
       "17                            ClinVar                         VUS   \n",
       "18                            ClinVar                      Benign   \n",
       "19                            ClinVar                         NaN   \n",
       "20                            ClinVar                         NaN   \n",
       "21                            ClinVar                         NaN   \n",
       "22                            ClinVar                         NaN   \n",
       "23                            ClinVar                         NaN   \n",
       "24                            ClinVar                         NaN   \n",
       "25                            ClinVar                         NaN   \n",
       "26                            ClinVar                         NaN   \n",
       "27                            ClinVar                         NaN   \n",
       "28                            ClinVar                         NaN   \n",
       "29                            ClinVar                         NaN   \n",
       "30                            ClinVar                         NaN   \n",
       "31                            ClinVar                         NaN   \n",
       "\n",
       "   Human disease reported in HGMD  \\\n",
       "0                             NaN   \n",
       "1                             NaN   \n",
       "2                  Breast cancer?   \n",
       "3                             NaN   \n",
       "4                             NaN   \n",
       "5                            FSGS   \n",
       "6                Frasier syndrome   \n",
       "7                            FSGS   \n",
       "8                Frasier syndrome   \n",
       "9                             NaN   \n",
       "10                            NaN   \n",
       "11               Frasier syndrome   \n",
       "12               Frasier syndrome   \n",
       "13               Frasier syndrome   \n",
       "14               Frasier syndrome   \n",
       "15                            NaN   \n",
       "16                            NaN   \n",
       "17                            NaN   \n",
       "18                            NaN   \n",
       "19  Ambiguous genetalia (ClinVar)   \n",
       "20                            NaN   \n",
       "21                            NaN   \n",
       "22                            NaN   \n",
       "23                            NaN   \n",
       "24                            NaN   \n",
       "25                            NaN   \n",
       "26                            NaN   \n",
       "27                            NaN   \n",
       "28                            NaN   \n",
       "29                            NaN   \n",
       "30                            NaN   \n",
       "31                            NaN   \n",
       "\n",
       "             In vitro evidence from other literatures  \\\n",
       "0                                                 NaN   \n",
       "1                                                 NaN   \n",
       "2                                                 NaN   \n",
       "3                                                 NaN   \n",
       "4                                                 NaN   \n",
       "5                                                 NaN   \n",
       "6                                                 NaN   \n",
       "7                                                 NaN   \n",
       "8   Loss of +KTS isoform in minigene assay and in ...   \n",
       "9                                                 NaN   \n",
       "10  in urine-derived patient cells shift from KTS+...   \n",
       "11             Loss of +KTS isoform in minigene assay   \n",
       "12             Loss of +KTS isoform in minigene assay   \n",
       "13             Loss of +KTS isoform in minigene assay   \n",
       "14                                                NaN   \n",
       "15                                                NaN   \n",
       "16                                                NaN   \n",
       "17                                                NaN   \n",
       "18                                                NaN   \n",
       "19                                                NaN   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "22                                                NaN   \n",
       "23                                                NaN   \n",
       "24                                                NaN   \n",
       "25                                                NaN   \n",
       "26                                                NaN   \n",
       "27                                                NaN   \n",
       "28                                                NaN   \n",
       "29                                                NaN   \n",
       "30                                                NaN   \n",
       "31                                                NaN   \n",
       "\n",
       "                                  Reference from HGMD  \\\n",
       "0                                                 NaN   \n",
       "1                                                 NaN   \n",
       "2                          Park (2018) BMC Cancer 18,   \n",
       "3                                                 NaN   \n",
       "4                                                 NaN   \n",
       "5             Kanemoto (2007) Pediatr Nephrol 22, 454   \n",
       "6                     Miyoshi (2006) Endocr J 53, 371   \n",
       "7         Gast (2016) Nephrol Dial Transplant 31, 961   \n",
       "8                   Kikuchi (1998) J Med Genet 35, 45   \n",
       "9                                                 NaN   \n",
       "10  https://www.kireports.org/article/S2468-0249(2...   \n",
       "11                   Barbaux (1997) Nat Genet 17, 467   \n",
       "12                   Bruening (1992) Nat Genet 1, 144   \n",
       "13                  Kikuchi (1998) J Med Genet 35, 45   \n",
       "14                   Barbaux (1997) Nat Genet 17, 467   \n",
       "15                                                NaN   \n",
       "16                                                NaN   \n",
       "17                                                NaN   \n",
       "18                                                NaN   \n",
       "19                                                NaN   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "22                                                NaN   \n",
       "23                                                NaN   \n",
       "24                                                NaN   \n",
       "25                                                NaN   \n",
       "26                                                NaN   \n",
       "27                                                NaN   \n",
       "28                                                NaN   \n",
       "29                                                NaN   \n",
       "30                                                NaN   \n",
       "31                                                NaN   \n",
       "\n",
       "                                                 Note  \n",
       "0                                                 NaN  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                                                 NaN  \n",
       "4                                                 NaN  \n",
       "5   Although not reported in ClinVar, it was repor...  \n",
       "6   Although not reported in ClinVar, it was repor...  \n",
       "7   Although not reported in ClinVar, it was repor...  \n",
       "8                                                 NaN  \n",
       "9                                                 NaN  \n",
       "10                                                NaN  \n",
       "11                                                NaN  \n",
       "12                                                NaN  \n",
       "13  Although not reported in ClinVar, it was repor...  \n",
       "14                                                NaN  \n",
       "15                                                NaN  \n",
       "16                                                NaN  \n",
       "17                                                NaN  \n",
       "18                                                NaN  \n",
       "19                                                NaN  \n",
       "20                                                NaN  \n",
       "21                                                NaN  \n",
       "22                                                NaN  \n",
       "23                                                NaN  \n",
       "24                                                NaN  \n",
       "25                                                NaN  \n",
       "26                                                NaN  \n",
       "27                                                NaN  \n",
       "28                                                NaN  \n",
       "29                                                NaN  \n",
       "30                                                NaN  \n",
       "31                                                NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e2f11987",
   "metadata": {},
   "outputs": [],
   "source": [
    "lit = lit.rename( columns = { 'cDNA ref': 'ref_c',\n",
    "                              'cDNA alt': 'alt_c',\n",
    "                              'cDNA_pos': 'hgvs_pos' } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6a7a8ebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "lit[ 'ref' ] = [ css.rev_complement( r ) for r in lit.ref_c ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c0dae4cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "lit[ 'alt' ] = [ css.rev_complement( a ) for a in lit.alt_c ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2464788f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ran this as a left merge bc one of the lit variants is too far into US intron (c.1355-87)\n",
    "idx_cols = [ 'hgvs_pos', 'hg19_pos', 'ref', 'alt', 'ref_c', 'alt_c' ]\n",
    "\n",
    "byvartbl_wide = byvartbl_wide.set_index( idx_cols ).merge( lit.set_index( idx_cols ),\n",
    "                                                           how = 'left',\n",
    "                                                           left_index = True,\n",
    "                                                           right_index = True ).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f6f88f71",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide[ 'clinvar' ] = byvartbl_wide[ 'ClinVar interp' ].notnull()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0beffd95",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide[ 'acmg' ] = byvartbl_wide[ 'Interpretation_ACMG_Varsome' ].notnull()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "60290195",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "( ( byvartbl_wide.clinvar ) | ( byvartbl_wide.acmg ) ).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "693f1e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide[ 'lit_interp' ] = [ 'Likely pathogenic/pathogenic (ClinVar)' if i == 'pathogenic' or i == 'likely pathogenic' \\\n",
    "                                  else 'VUS/Conflicting' if i == 'VUS' or i == 'conflicting interpretations' \\\n",
    "                                  else 'Likely benign/benign' if i == 'benign' or i == 'likely benign' \\\n",
    "                                  else '' for i in byvartbl_wide[ 'ClinVar interp' ] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cffce992",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide.loc[ ( byvartbl_wide.Interpretation_ACMG_Varsome == 'Pathogenic' ) & ( byvartbl_wide.lit_interp == '' ), 'lit_interp' ] = 'Likely pathogenic/pathogenic (Literature)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "55d5384b",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide[ 'KTS_+_ratio_log2' ] = np.log2( byvartbl_wide[ 'wmean_KTS+'] / byvartbl_wide[ 'wmean_KTS-' ] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a7d35a25",
   "metadata": {},
   "outputs": [],
   "source": [
    "byvartbl_wide.to_csv( data_in + 'wt1_ex9_by_var_effects_snvs_wide-2022-0727b.txt',\n",
    "                      sep = '\\t',\n",
    "                      index = False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e775c0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
